interpreters/wamr/wamr_custom_init.c

```diff
@@ -0,0 +1,657 @@
+  if (apv == NULL)
+    {
+      return;
+    }
+  while (*apv != 0)
+    {
+      *apv = (uintptr_t)addr_native_to_app((void *)*apv);
+      apv++;
+    }
+}
+
+static pthread_mutex_t g_compare_mutex = PTHREAD_MUTEX_INITIALIZER;
+static wasm_exec_env_t g_compare_env;
+static void           *g_compare_func;
+
+static int
+compare_proxy(const void *a, const void *b)
+{
+  wasm_module_inst_t module_inst = get_module_inst(g_compare_env);
+  uint32_t argv[2];
+
+  argv[0] = addr_native_to_app((void *)a);
+  argv[1] = addr_native_to_app((void *)b);
+
+  return wasm_runtime_call_indirect(g_compare_env,
+           (uint32_t)addr_native_to_app(g_compare_func), 2, argv) ?
+             argv[0] : 0;
+}
+
+#ifndef GLUE_FUNCTION_qsort
+#define GLUE_FUNCTION_qsort
+void glue_qsort(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2,
+                uintptr_t parm3, uintptr_t parm4)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  pthread_mutex_lock(&g_compare_mutex);
+  g_compare_env = env;
+  g_compare_func = parm4;
+  qsort((FAR void *)parm1, (size_t)parm2,
+        (size_t)parm3, compare_proxy);
+  pthread_mutex_unlock(&g_compare_mutex);
+}
+
+#endif /* GLUE_FUNCTION_qsort */
+
+#ifndef GLUE_FUNCTION_bsearch
+#define GLUE_FUNCTION_bsearch
+uintptr_t glue_bsearch(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2,
+                       uintptr_t parm3, uintptr_t parm4, uintptr_t parm5)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  uintptr_t ret;
+
+  pthread_mutex_lock(&g_compare_mutex);
+  g_compare_env = env;
+  g_compare_func = parm5;
+  ret = bsearch((FAR const void *)parm1,
+                (FAR const void *)parm2,
+                (size_t)parm3, (size_t)parm4, compare_proxy);
+  pthread_mutex_unlock(&g_compare_mutex);
+  ret = addr_native_to_app((void *)ret);
+  return ret;
+}
+
+#endif /* GLUE_FUNCTION_bsearch */
+
+static void glue_msghdr_begin(wasm_module_inst_t module_inst,
+                              FAR struct msghdr *hdr)
+{
+  int i;
+
+  hdr->msg_iov = addr_app_to_native((uintptr_t)hdr->msg_iov);
+
+  for (i = 0; i < hdr->msg_iovlen; i++)
+    {
+      hdr->msg_iov[i].iov_base =
+        addr_app_to_native((uintptr_t)hdr->msg_iov[i].iov_base);
+    }
+
+  if (hdr->msg_name != NULL && hdr->msg_namelen > 0)
+    {
+      hdr->msg_name = addr_app_to_native(hdr->msg_name);
+    }
+
+  if (hdr->msg_control != NULL && hdr->msg_controllen > 0)
+    {
+      hdr->msg_control = addr_app_to_native((uintptr_t)hdr->msg_control);
+    }
+}
+
+static void glue_msghdr_end(wasm_module_inst_t module_inst,
+                            FAR struct msghdr *hdr)
+{
+  int i;
+
+  for (i = 0; i < hdr->msg_iovlen; i++)
+    {
+      hdr->msg_iov[i].iov_base =
+        addr_native_to_app((uintptr_t)hdr->msg_iov[i].iov_base);
+    }
+
+  hdr->msg_iov = addr_native_to_app((uintptr_t)hdr->msg_iov);
+
+  if (hdr->msg_name != NULL && hdr->msg_namelen > 0)
+    {
+      hdr->msg_name = addr_native_to_app(hdr->msg_name);
```

Comment:
are you assuming host pointer size?

Reply:
It might be a issue since the host pointer maybe 8 byte size, the pointer size is 4 byte, that means on 64 bit platform, the offset (host pointer - start address of linear memory of wasm) may exceed the 32 bit len.